core/src/main/java/org/eclipse/hono/util/ResourceLimitsPeriod.java

```diff
@@ -0,0 +1,172 @@
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.time.OffsetDateTime;
+import java.time.YearMonth;
+import java.time.temporal.ChronoUnit;
+import java.util.Objects;
+
+/**
+ * The period definition corresponding to a resource limit for a tenant.
+ */
+@JsonInclude(JsonInclude.Include.NON_DEFAULT)
+public class ResourceLimitsPeriod {
+
+    @JsonProperty(value = TenantConstants.FIELD_PERIOD_MODE, required = true)
+    private String mode;
+
+    @JsonProperty(value = TenantConstants.FIELD_PERIOD_NO_OF_DAYS)
+    private int noOfDays;
+
+    /**
+     * Gets the mode of period for resource limit calculation.
+     *
+     * @return The mode of period for resource limit calculation.
+     */
+    public final String getMode() {
+        return mode;
+    }
+
+    /**
+     * Sets the mode of period for resource limit calculation.
+     *
+     * @param mode The mode of period for resource limit calculation.
+     * @return  a reference to this for fluent use.
+     * @throws NullPointerException if mode is {@code null}.
+     */
+    public final ResourceLimitsPeriod setMode(final String mode) {
+        this.mode = Objects.requireNonNull(mode);
+        return this;
+    }
+
+    /**
+     * Gets the number of days for which resource usage is calculated.
+     *
+     * @return The number of days for a resource limit calculation.
+     */
+    public final int getNoOfDays() {
+        return noOfDays;
+    }
+
+    /**
+     * Sets the number of days for which resource usage is calculated.
+     *
+     * @param noOfDays The number of days for which resource usage is calculated.
+     * @return  a reference to this for fluent use.
+     * @throws IllegalArgumentException if the number of days is negative.
+     */
+    public final ResourceLimitsPeriod setNoOfDays(final int noOfDays) {
+        if (noOfDays < 0) {
+            throw new IllegalArgumentException("Number of days property must be  set to value >= 0");
+        }
+        this.noOfDays = noOfDays;
+        return this;
+    }
+
+    /**
+     * Calculates the period in days for which the resource usage like volume of used data, 
+     * connection duration etc. is to be calculated based on the mode defined by 
+     * {@link TenantConstants#FIELD_PERIOD_MODE}.
+     *
+     * @param effectiveSince The point of time on which the resource limit came into effect.
+     * @param targetDateTime The target date and time used for the resource usage period calculation.
+     * @param mode The mode of the period defined by {@link TenantConstants#FIELD_PERIOD_MODE}.
+     * @param periodInDays The number of days defined by {@link TenantConstants#FIELD_PERIOD_NO_OF_DAYS}. 
+     * @return The period in days for which the resource usage is to be calculated.
+     * @throws NullPointerException if effectiveSince, targetDateTime or mode is {@code null}.
+     */
+    public static long calculateResourceUsagePeriod(
```

Comment:
I would have expected this method to not be final but instead operate on the instance's fields instead of passing in the mode and periodInDays or doesn't that work?

Reply:
The `DataVolume` is still using `DataVolumePeriod` not  `ResourceLimitsPeriod`. Until then the mode and periodInDays have to be passed.

Reply:
We could create a common abstract base class for `DataVolumePeriod` and `ResourceLimitsPeriod` which defines the shared functionality. WDYT?

Reply:
Yes we can create one base class. But I think then it needs to be marked as _deprecated_ for removal in 2.0.0. Once the _DataVolumePeriod_ has been removed, then this new base class should be merged with `ResourceLimitsPeriod`. Considering the new base class and the related changes, I am wondering if it is worthwhile to move now the `calculateResourceUsagePeriod` from the `PrometheusBasedResourceLimitChecks`. I think while removing the `DataVolumePeriod`, this activity can also be carried out. WDYT?

Reply:
I agree, it's not worth the effort to move the code at all for the time being ...

Reply:
For the time being I have reverted this method back to the `PrometheusBasedResourceLimitChecks`. Please take a look.